### ISSUE TITLE
DEFAULT phase transition occurs too late for mixing some core game classes

### ISSUE THREAD

**Issue by Zidane** [SpongePowered/Mixin#67]
One thing that we've hit in SpongeForge is the fact that we are injecting into GameData#registerBlock(Block, String, int);int and therefore we can't remap this as the method name is obviously not obfuscated BUT the Block parameter needs to be. Sooo

* We do remap=false but at production the injection fails as it expects registerBlock(atr, String, int);int where atr = Block.class
* we do remap = true but at compile-time the mixin AP fails because it finds no obfuscation mapping for registerBlock (as there wouldn't be one since it isn't obfuscated at all).

With the latter option, I tried shoehoring an entry into extraSrg that basically just maps registerBlock back onto the same thing with:

MD: net/minecraftforge/fml/common/registry/GameData/registerBlock (Lnet/minecraft/block/Block;Ljava/lang/String;I)I net/minecraftforge/fml/common/registry/GameData/registerBlock (Lnet/minecraft/block/Block;Ljava/lang/String;I)I

Unless that mapping is incorrect...

We need guidance or a fix for this ASAP.

**Comment by Zidane** [comment]
Tagging a few people:

@bloodmc @Aaron1011 @gabizou

**Comment by Zidane** [comment]
After more investigation, this is happening before we need to mixin to GameData and that can only be done in MixinEnvironment.PREINIT. At this point, this is before FML's deobf transformer has the chance to deobfuscate the class files.

Here is the error you'll get with Sponge's feature/traits branch and just running it:

```
C:\Users\csanders\Desktop\New folder>java -jar sponge-1.8-1499-2.1DEV-0.jar go

Sponge v1.8-1499-2.1DEV-0+unknown-b0.git-unknown for Minecraft Forge 1499
   implementing SpongeAPI version 2.1-SNAPSHOT
:: resolving dependencies :: spongepowered#spongeruntime;working@MCS-ZIDANE-LPT
        confs: [libraries]
        found net.minecraft#minecraft_server;1.8 in minecraft
        found net.minecraftforge#forge#universal;1.8-11.14.3.1499 in forge
        found net.minecraft#launchwrapper;1.9 in mojang
        found org.ow2.asm#asm-all;5.0.3 in mojang
        found com.typesafe.akka#akka-actor_2.11;2.3.3 in spongemaven
        found com.typesafe#config;1.2.1 in spongemaven
        found org.scala-lang#scala-actors-migration_2.11;1.1.0 in spongemaven
        found org.scala-lang#scala-compiler;2.11.1 in spongemaven
        found org.scala-lang.plugins#scala-continuations-library_2.11;1.0.2 in spongemaven
        found org.scala-lang.plugins#scala-continuations-plugin_2.11.1;1.0.2 in spongemaven
        found org.scala-lang#scala-library;2.11.1 in spongemaven
        found org.scala-lang.modules#scala-parser-combinators_2.11;1.0.1 in spongemaven
        found org.scala-lang#scala-reflect;2.11.1 in spongemaven
        found org.scala-lang.modules#scala-swing_2.11;1.0.1 in spongemaven
        found org.scala-lang.modules#scala-xml_2.11;1.0.2 in spongemaven
        found lzma#lzma;0.0.1 in mojang
        found net.sf.jopt-simple#jopt-simple;4.5 in mojang
        found com.ibm.icu#icu4j-core-mojang;51.2 in mojang
:: resolution report :: resolve 304ms :: artifacts dl 15ms
        ---------------------------------------------------------------------
        |                  |            modules            ||   artifacts   |
        |       conf       | number| search|dwnlded|evicted|| number|dwnlded|
        ---------------------------------------------------------------------
        |     libraries    |   18  |   0   |   0   |   0   ||   18  |   0   |
        ---------------------------------------------------------------------
:: retrieving :: spongepowered#spongeruntime
        confs: [libraries]
        0 artifacts copied, 18 already retrieved (0kB/19ms)
[14:11:36] [main/INFO]: Loading tweak class name net.minecraftforge.fml.common.launcher.FMLServerTweaker
[14:11:36] [main/INFO]: Using primary tweak class name net.minecraftforge.fml.common.launcher.FMLServerTweaker
[14:11:36] [main/INFO]: Calling tweak class net.minecraftforge.fml.common.launcher.FMLServerTweaker
[14:11:36] [main/INFO]: Forge Mod Loader version 8.99.196.1499 for Minecraft 1.8 loading
[14:11:36] [main/INFO]: Java is Java HotSpot(TM) 64-Bit Server VM, version 1.8.0_51, running on Windows 8.1:amd64:6.3, installed at C:\Program Files\Java\jre1.8.0_51
[14:11:36] [main/INFO]: Found a command line coremod : org.spongepowered.mod.SpongeCoremod
[14:11:38] [main/INFO]: SpongePowered MIXIN Subsystem Version=0.4.4 Source=file:/C:/Users/csanders/Desktop/New%20folder/sponge-1.8-1499-2.1DEV-0.jar Env=SERVER
[14:11:38] [main/INFO]: Adding new token provider org.spongepowered.mod.SpongeCoremod$TokenProvider to MixinEnvironment[PREINIT]
[14:11:38] [main/INFO]: Adding new token provider org.spongepowered.mod.SpongeCoremod$TokenProvider to MixinEnvironment[DEFAULT]
[14:11:38] [main/INFO]: Loading tweak class name net.minecraftforge.fml.common.launcher.FMLInjectionAndSortingTweaker
[14:11:38] [main/INFO]: Loading tweak class name org.spongepowered.asm.mixin.MixinEnvironment$EnvironmentStateTweaker
[14:11:38] [main/INFO]: Loading tweak class name net.minecraftforge.fml.common.launcher.FMLDeobfTweaker
[14:11:38] [main/INFO]: Calling tweak class net.minecraftforge.fml.common.launcher.FMLInjectionAndSortingTweaker
[14:11:38] [main/INFO]: Calling tweak class net.minecraftforge.fml.common.launcher.FMLInjectionAndSortingTweaker
[14:11:38] [main/INFO]: Calling tweak class net.minecraftforge.fml.relauncher.CoreModManager$FMLPluginWrapper
[14:11:39] [main/INFO]: Found valid fingerprint for Minecraft Forge. Certificate fingerprint e3c3d50c7c986df74c645c0ac54639741c90a557
[14:11:39] [main/INFO]: Calling tweak class net.minecraftforge.fml.relauncher.CoreModManager$FMLPluginWrapper
[14:11:39] [main/INFO]: Calling tweak class net.minecraftforge.fml.relauncher.CoreModManager$FMLPluginWrapper
[14:11:39] [main/INFO]: Calling tweak class org.spongepowered.asm.mixin.MixinEnvironment$EnvironmentStateTweaker
[14:11:39] [main/INFO]: Calling tweak class net.minecraftforge.fml.common.launcher.FMLDeobfTweaker
[14:11:39] [main/FATAL]: Mixin failed applying mixins.forge.base.json:fml.common.registry.MixinGameData -> net.minecraftforge.fml.common.registry.GameData: org.spongepowered.asm.mixin.injection.InvalidInjectionException Invalid descriptor on callback: expected (Latr;Ljava/lang/String;Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;)V but found (Lnet/minecraft/block/Block;Ljava/lang/String;ILorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;)V
org.spongepowered.asm.mixin.injection.InvalidInjectionException: Invalid descriptor on callback: expected (Latr;Ljava/lang/String;Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;)V but found (Lnet/minecraft/block/Block;Ljava/lang/String;ILorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;)V
        at org.spongepowered.asm.mixin.injection.callback.CallbackInjector.inject(CallbackInjector.java:403) ~[sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.asm.mixin.injection.callback.CallbackInjector.inject(CallbackInjector.java:340) ~[sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.asm.mixin.injection.code.Injector.injectInto(Injector.java:108) ~[sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.asm.mixin.injection.struct.InjectionInfo.inject(InjectionInfo.java:184) ~[sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.asm.mixin.transformer.MixinApplicator.applyInjections(MixinApplicator.java:705) ~[sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.asm.mixin.transformer.MixinApplicator.applyMixin(MixinApplicator.java:208) ~[sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.asm.mixin.transformer.MixinApplicator.apply(MixinApplicator.java:189) ~[sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.asm.mixin.transformer.MixinTransformer.apply(MixinTransformer.java:594) ~[sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.asm.mixin.transformer.MixinTransformer.applyMixins(MixinTransformer.java:566) [sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.asm.mixin.transformer.MixinTransformer.transform(MixinTransformer.java:402) [sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.asm.mixin.transformer.MixinTransformer$Proxy.transform(MixinTransformer.java:104) [sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:276) [launchwrapper-1.9.jar:?]
        at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:174) [launchwrapper-1.9.jar:?]
        at java.lang.ClassLoader.loadClass(Unknown Source) [?:1.8.0_51]
        at java.lang.ClassLoader.loadClass(Unknown Source) [?:1.8.0_51]
        at java.lang.Class.getDeclaredMethods0(Native Method) [?:1.8.0_51]
        at java.lang.Class.privateGetDeclaredMethods(Unknown Source) [?:1.8.0_51]
        at java.lang.Class.privateGetMethodRecursive(Unknown Source) [?:1.8.0_51]
        at java.lang.Class.getMethod0(Unknown Source) [?:1.8.0_51]
        at java.lang.Class.getMethod(Unknown Source) [?:1.8.0_51]
        at net.minecraftforge.fml.common.launcher.FMLDeobfTweaker.injectIntoClassLoader(FMLDeobfTweaker.java:39) [forge-1.8-11.14.3.1499.jar:?]
        at net.minecraft.launchwrapper.Launch.launch(Launch.java:114) [launchwrapper-1.9.jar:?]
        at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.9.jar:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_51]
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_51]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_51]
        at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_51]
        at net.minecraftforge.fml.relauncher.ServerLaunchWrapper.run(ServerLaunchWrapper.java:43) [forge-1.8-11.14.3.1499.jar:?]
        at net.minecraftforge.fml.relauncher.ServerLaunchWrapper.main(ServerLaunchWrapper.java:12) [forge-1.8-11.14.3.1499.jar:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_51]
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_51]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_51]
        at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_51]
        at org.spongepowered.launch.handlers.RunHandler.run(RunHandler.java:84) [sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.launch.handlers.GoHandler.run(GoHandler.java:51) [sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.launch.Main.dispatch(Main.java:134) [sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
        at org.spongepowered.launch.Main.main(Main.java:82) [sponge-1.8-1499-2.1DEV-0.jar:1.8-1499-2.1DEV-0+unknown-b0.git-unknown]
A problem occurred running the Server launcher.[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]: java.lang.reflect.InvocationTargetException
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at java.lang.reflect.Method.invoke(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at net.minecraftforge.fml.relauncher.ServerLaunchWrapper.run(ServerLaunchWrapper.java:43)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at net.minecraftforge.fml.relauncher.ServerLaunchWrapper.main(ServerLaunchWrapper.java:12)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at java.lang.reflect.Method.invoke(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.launch.handlers.RunHandler.run(RunHandler.java:84)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.launch.handlers.GoHandler.run(GoHandler.java:51)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.launch.Main.dispatch(Main.java:134)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.launch.Main.main(Main.java:82)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]: Caused by: java.lang.NoClassDefFoundError: net/minecraftforge/fml/common/registry/GameData
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at java.lang.Class.getDeclaredMethods0(Native Method)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at java.lang.Class.privateGetDeclaredMethods(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at java.lang.Class.privateGetMethodRecursive(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at java.lang.Class.getMethod0(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at java.lang.Class.getMethod(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at net.minecraftforge.fml.common.launcher.FMLDeobfTweaker.injectIntoClassLoader(FMLDeobfTweaker.java:39)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at net.minecraft.launchwrapper.Launch.launch(Launch.java:114)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    ... 14 more
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]: Caused by: java.lang.ClassNotFoundException: net.minecraftforge.fml.common.registry.GameData
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:188)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at java.lang.ClassLoader.loadClass(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at java.lang.ClassLoader.loadClass(Unknown Source)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    ... 22 more
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]: Caused by: org.spongepowered.asm.mixin.MixinApplyError: Mixin [mixins.forge.base.json:fml.common.registry.MixinGameData] from phase [PREINIT] in config [mixins.forge.base.json] FAILED
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.transformer.MixinTransformer.handleMixinErrorState(MixinTransformer.java:634)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.transformer.MixinTransformer.transform(MixinTransformer.java:408)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.transformer.MixinTransformer$Proxy.transform(MixinTransformer.java:104)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at net.minecraft.launchwrapper.LaunchClassLoader.runTransformers(LaunchClassLoader.java:276)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:174)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    ... 24 more
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]: Caused by: org.spongepowered.asm.mixin.injection.InvalidInjectionException: Invalid descriptor on callback: expected (Latr;Ljava/lang/String;Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;)V but found (Lnet/minecraft/block/Block;Ljava/lang/String;ILorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;)V
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.injection.callback.CallbackInjector.inject(CallbackInjector.java:403)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.injection.callback.CallbackInjector.inject(CallbackInjector.java:340)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.injection.code.Injector.injectInto(Injector.java:108)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.injection.struct.InjectionInfo.inject(InjectionInfo.java:184)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.transformer.MixinApplicator.applyInjections(MixinApplicator.java:705)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.transformer.MixinApplicator.applyMixin(MixinApplicator.java:208)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.transformer.MixinApplicator.apply(MixinApplicator.java:189)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.transformer.MixinTransformer.apply(MixinTransformer.java:594)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.transformer.MixinTransformer.applyMixins(MixinTransformer.java:566)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    at org.spongepowered.asm.mixin.transformer.MixinTransformer.transform(MixinTransformer.java:402)
[14:11:39] [main/INFO]: [java.lang.Throwable$WrappedPrintStream:println:-1]:    ... 27 more
```

**Comment by Mumfrey** [comment]
This isn't too bad to deal with I don't think. Basically remap is there in the same way that `@Override` and other markers appear in code to be a *hint* to the AP that "hey, this isn't remappable, don't panic if you can't find an obf mapping for it". Of course in the real reobf pass, all hard tokens like class names get remapped as part of the AST and it doesn't matter if they don't appear in the obfuscation tables because class names get remapped automagically.

Of course the AP doesn't have this info, so what I really need to add is a failover for `remap=false` where it still attemtps to remap class name tokens in the signature and generate the srgs where appropriate. I don't think this should be super tricky since for the most part these mappings are readily available in the srg file that's passed to the AP (at least from memory).

So ya, the correct approach is still to mark them as `remap=false`, I will simply add handling in the AP at that point for remapping the class name tokens in the signature.

**Comment by Zidane** [comment]
I'm not sure if I have adequately explained the issue wither @Mumfrey . Can you come on to TS? @Aaron1011 knows what is going on.

**Comment by bloodmc** [comment]
As Zidane stated, by the time the CallbackInjector attempts to inject the mixin into the target class, GameData has not been reobfuscated and is still using notch names. This causes the mixin method in our MixinGameData class to fail to find a proper match since our mixin MethodNode descriptor contains searge and not notch.

**Comment by Mumfrey** [comment]
You cannot load game classes of any kind in `PREINIT`. I think I've just realised why this issue is actually invalid for Sponge (it's valid for my cases, like in LiteLoader but not for a Forge environment). I'll try to get on TS tonight to talk about it because I think you're probably trying to do something that's impossible.

**Comment by Mumfrey** [comment]
EDIT2: I think the issue here is actually timing. Reading through [FMLDeobfTweaker](https://github.com/MinecraftForge/MinecraftForge/blob/master/src/main/java/net/minecraftforge/fml/common/launcher/FMLDeobfTweaker.java), basically what I consider `PREINIT` is complete as of [line 35](https://github.com/MinecraftForge/MinecraftForge/blob/master/src/main/java/net/minecraftforge/fml/common/launcher/FMLDeobfTweaker.java#L35), at this point we have transitioned into a state between `PREINIT` and `DEFAULT` where it's **sort of** safe to load game classes (though probably unsafe to load any that wouldn't otherwise be loaded at this point). The deobf tweaker duly [initialises the Loader](https://github.com/MinecraftForge/MinecraftForge/blob/master/src/main/java/net/minecraftforge/fml/common/launcher/FMLDeobfTweaker.java#L37-L43) which is where the GameData class is being indirectly loaded. However because we now need to be acting like we're in `DEFAULT` but as far as the transformer is concerned we're still in `PREINIT`, which is a problem.

So what the issue actually seems to boil down to is that we have an interim state *between* `PREINIT` and `DEFAULT` where **some** game classes are caused to be loaded, the Deobf Transformer is available but isn't being run on incoming mixin code because my transformer hasn't *realised* that the transition has occurred.

So the AP stuff is a red herring, that's working as it should. Where the issue actually lies appears to be in the transformer and this temporary state that exists between `PREINIT` and `DEFAULT` that I hadn't accounted for.

**Comment by Zidane** [comment]
@Mumfrey 

I came to that realization last night as well. I thought a potential fix is to accept that PREINIT cannot touch SRG classes and make a new state between INIT and DEFAULT that would be able to acquire the SRG name now available from the FMLDeobfTweaker but still be early enough to mix it. 

I am not sure the correct spot to put this at...

**Comment by Mumfrey** [comment]
@Zidane yes this is absolutely the case, it is an absolutely unbreachable rule that **no obfuscated classes can be referenced during PREINIT**, this is by design and bad things will happen if this rule is ever broken. However as you say, what we need is a new phase *between* `PREINIT` and `DEFAULT` and this interestingly dovetails into the need to [add support for arbitrary phases in the future](https://github.com/SpongePowered/Mixin/issues/57#issuecomment-119211503) which I have already mentioned in the past I intend to support. However in the short term I think a new phase `INIT` can be added which reflects this `post-preinit-but-after-deobf-init-and-before-default` phase of startup. In fact it's crucial that this phase is recognised because it's an important step in the startup process which has hitherto not been appreciated.

Adding arbitrary phases of course depends on being able to define criteria for those phases, and so step one is identifying how we will spot the transition and trigger that phase change. As a starter for 10 I would propose looking for the classload of the deobf transformer, or monitoring the transformer list for the addition of a specific transformer.

Updating the title of this issue to be more reflective of the actual problem.

**Comment by Zidane** [comment]
https://github.com/MinecraftForge/MinecraftForge/blob/master/src/main/java/net/minecraftforge/fml/common/launcher/FMLDeobfTweaker.java#L26

So from my investigation, this registers and loads the transformer class with Launch but unfortunately the rest of injectIntoClassLoader will also start initializing the game which will include GameData.

It would be a simple affair to add an "INIT" phase right after registerTransfomer if...

A. Launch provided any callback for when a transformer is registered
or
B. We could make changes to FML to call our INIT phase.

From what I can see, unless you have some hocus pocus @Mumfrey , there is no way to tell that Deobf is being registered at the time necessary to do this.

So if I am indeed right, alternative less desirable solutions are needed. One that comes to mind is allowing a mixin config to specify a "fallback" SRG to use in-case the CallbackInjector fails to find the target. Basically if normal lookup fails to find a class, use fallback SRG, remap the name, try again...if you can't find the target then throw the exception you are doing currently. 

Its a seriously nasty hack but it would work.

**Comment by Zidane** [comment]
I've even dreamed up completely outlandish solutions such as adding a mixin via PREINIT to mix CoreModManager getAccessTransformers at HEAD in-which would check if the deobf transformer is there and if so...transition to INIT

**Comment by Mumfrey** [comment]
Naw this is trivial, fear not. See the crucial point to figure out when a transition \*must\* occur is before the point that the transition is actually *handled*, which in this case is at the entry point to the transformer (the actual transformer not the proxy). See what happens currently is that the transition is detected and the "current" environment set inside [what is a tweaker callback](https://github.com/SpongePowered/Mixin/blob/master/src/main/java/org/spongepowered/asm/mixin/MixinEnvironment.java#L348), however this transition isn't actually \*acted upon\* [until the next transformation](https://github.com/SpongePowered/Mixin/blob/master/src/main/java/org/spongepowered/asm/mixin/transformer/MixinTransformer.java#L358), when the transformer is invoked.

Basically, since this is the only point where we **care** about the transition, it's also the perfect point to **check** for the transition, because let's face it - nothing else actually matters except that we're in the correct environment when mixing in.

So basically, it's not too tricky to do, it might be a little *slower* than before (by an infinitesimal amount), but it allows pretty arbitrary transitions to be detected at pretty much any point which matters.

**Comment by Mumfrey** [comment]
> I've even dreamed up completely outlandish solutions such as adding a mixin via PREINIT to mix CoreModManager getAccessTransformers at HEAD in-which would check if the deobf transformer is there and if so...transition to INIT

Also 100% feasible, there are lots of ways of handling this though I think my solution is probably simpler.

**Comment by Zidane** [comment]
@Mumfrey 

I see, in transform I could check to see if the current environment is PREINIT and Launch has the DeoobfTransformer and PREINIT is initialized...if so transition to INIT. This would happen before you init the environment.

Correct? If so, I could do this I think.

**Comment by JBYoshi** [comment]
>> I've even dreamed up completely outlandish solutions such as adding a mixin via PREINIT to mix CoreModManager getAccessTransformers at HEAD in-which would check if the deobf transformer is there and if so...transition to INIT

> Also 100% feasible, there are lots of ways of handling this though I think my solution is probably simpler.

I don't think this would work in production, as `CoreModManager` would already be loaded (it would be loading the tweaker).

**Comment by Zidane** [comment]
@JBYoshi 

MixinTweaker is added via Mixin to Launch.

**Comment by Zidane** [comment]
@Mumfrey 

After even more investigation and writing some code, I've came to a conclusion on a few things.

I've added INIT phase to Mixin and have that loading the mixin file for its phase. That part was dead simple. I also know the earliest points for switching to init but that is where things get difficult.

The problem is the caching of the transformer delegates that happen when an environment are initialized. The appropriate spot to goToPhase(INIT) is within MixinTransformer#transform as Mixin isn't a coremod (and therefore won't receive the "callback" getAccessTransformer from FML) and we must get the DeobfTransformer in the cached transformer list of the INIT environment phase. Unfortunately, at that point, the transformer list is already being iterated by LaunchClassLoader and ends up in dying from a CME.

At the least, I've confirmed this would work (besides the exception) as I now see mapped SRG class names for INIT mixins in production. I'm just not sure how to tackle getting around that issue.

**Comment by Mumfrey** [comment]
> I've added INIT phase to Mixin and have that loading the mixin file for its phase. That part was dead simple. I also know the earliest points for switching to init but that is where things get difficult.

That's cool, and not surprising as that was entirely set up to be extensible.

> The problem is the caching of the transformer delegates that happen when an environment are initialized. The appropriate spot to goToPhase(INIT) is within MixinTransformer#transform as Mixin isn't a coremod (and therefore won't receive the "callback" getAccessTransformer from FML) and we must get the DeobfTransformer in the cached transformer list of the INIT environment phase. Unfortunately, at that point, the transformer list is already being iterated by LaunchClassLoader and ends up in dying from a CME.

Well there are two possibilities. One, we can mark (for now) environment transitions as either needing a proxy update or not, and just outright saying that for INIT we don't bother adding a new proxy. Since this whole thing is a quick and dirty hack and will be replaced I don't have a massive issue with that. The other option is to look for a transition point which is *not* within the transformer loop.

A hacky but viable option would be to get fml's `FMLRelaunchLog` inner `Logger` instance (which can be done by `FMLRelaunchLog.log.getLogger()` or just a direct `LogManager.getLogger("FML")`) and add a new `Appender` to it (will need to cast down to the class but that's not tricky).

Once you can add an `Appender`, you simply write a skeleton implementation of the Appender with:

```java
@Override public void append(LogEvent event) {
    if (event.getLevel() == Level.DEBUG && "Validating minecraft".equals(event.getMessage().getFormat())) {
        // transition to INIT
    }
}
```

This basically allows us to watch for [this line](https://github.com/MinecraftForge/MinecraftForge/blob/master/src/main/java/net/minecraftforge/fml/common/launcher/FMLDeobfTweaker.java#L37) and safely make the transition because the transformer list isn't being iterated at that point.

> At the least, I've confirmed this would work (besides the exception) as I now see mapped SRG class names for INIT mixins in production. I'm just not sure how to tackle getting around that issue.

That's good, I expected as much to be honest because reading your stack trace above nicely highlighted the problem (thanks to FMLDeobfTweaker appearing in the stack trace).

Bear in mind this is just a temporary fix so you can basically do whatever you want and we'll replace it with something more sensible when time allows. If you commit your changes on a branch I'll just set up my jenkins server to build that branch. Bump the version to 0.4.6 on the branch as well when you do so.

**Comment by Zidane** [comment]
Ha I love that. What an awesome but dirty hack :D. I'll give it a whirl tonight.

**Comment by Mumfrey** [comment]
I might not work but it should, I actually do a similar thing and [programmatically inject a new custom appender](https://www.assembla.com/spaces/liteloader/subversion/source/HEAD/LiteLoader/trunk/java/common/com/mumfrey/liteloader/util/log/LiteLoaderLogger.java#ln65) in liteloader in order to be able to watch messages being logged so I can display the log tail ingame. Notice the usefulness of extending from `AbstractAppender` instead of implementing the `Appender` interface directly.

**Comment by Zidane** [comment]
@Mumfrey 

Seems like I just can't catch a break. So the appender works great in dev but in production while my appender is constructed (my printout shows this), I never get the log record for "Validating Minecraft".

**Comment by Zidane** [comment]
@Mumfrey 

I've pushed a solution to the initphase branch that I've confirmed works across dev and production (with ivy launch as well).

If that hack is something you can stomach for now, we really need to get this built and available on the repo ASAP so Sponge can continue development.

**Comment by Mumfrey** [comment]
My Jenkins server [is now set up to build and push that branch](http://jenkins.liteloader.com/job/Mixin/) automatically.

**Comment by Mumfrey** [comment]
Closing in favour of #78
